crates/database/src/order_execution.rs

```diff
@@ -1,33 +1,91 @@
 use {
-    crate::{auction::AuctionId, OrderUid},
+    crate::{auction::AuctionId, Address, OrderUid},
     bigdecimal::BigDecimal,
-    sqlx::PgConnection,
+    sqlx::{
+        postgres::{PgHasArrayType, PgTypeInfo},
+        PgConnection,
+    },
+    std::collections::HashMap,
 };
 
+#[derive(Clone, Debug, Eq, PartialEq, sqlx::Type, sqlx::FromRow)]
+pub struct FeeAsset {
+    pub amount: BigDecimal,
+    pub token: Address,
+}
+
+// explains that the equivalent Postgres type is already defined in the database
+impl PgHasArrayType for FeeAsset {
+    fn array_type_info() -> PgTypeInfo {
+        PgTypeInfo::with_name("_feeasset")
+    }
+}
+
 pub async fn save(
     ex: &mut PgConnection,
     order: &OrderUid,
     auction: AuctionId,
     block_number: i64,
     executed_fee: &BigDecimal,
+    executed_protocol_fees: Option<&[FeeAsset]>,
 ) -> Result<(), sqlx::Error> {
     const QUERY: &str = r#"
-INSERT INTO order_execution (order_uid, auction_id, reward, surplus_fee, block_number)
-VALUES ($1, $2, $3, $4, $5)
+INSERT INTO order_execution (order_uid, auction_id, reward, surplus_fee, block_number, protocol_fees)
+VALUES ($1, $2, $3, $4, $5, $6)
 ON CONFLICT (order_uid, auction_id)
-DO UPDATE SET reward = $3, surplus_fee = $4, block_number = $5
+DO UPDATE SET reward = $3, surplus_fee = $4, block_number = $5, protocol_fees = $6
 ;"#;
     sqlx::query(QUERY)
         .bind(order)
         .bind(auction)
         .bind(0.) // reward is deprecated but saved for historical analysis
         .bind(Some(executed_fee))
         .bind(block_number)
+        .bind(executed_protocol_fees)
         .execute(ex)
         .await?;
     Ok(())
 }
 
+// fetch protocol fees for all keys in the filter
+pub async fn executed_protocol_fees(
+    ex: &mut PgConnection,
+    keys_filter: &[(AuctionId, OrderUid)],
+) -> Result<HashMap<(AuctionId, OrderUid), Vec<FeeAsset>>, sqlx::Error> {
+    if keys_filter.is_empty() {
+        return Ok(HashMap::new());
+    }
+
+    let mut fees = HashMap::new();
+    for (auction_id, order_uid) in keys_filter {
```

Comment:
maybe is worth having a query that fetches many from an slice of `(auction_id, order_uid)` for faster querying? :thinking:

Reply:
Tried already with query builder but after spending quite some time trying to make it work with composite type in postgresql and failing, i just gave up

Reply:
Yes, please, this should only be one roundtrip (otherwise exposing a batching API is just confusing the caller).

I don't think the composite type is helping is much, so how we just remove it in favour of two vanilla columns?

Reply:
> I don't think the composite type is helping is much, so how we just remove it in favour of two vanilla columns?

AFAIU the alternative would be an extra table with: `order_uid`, `auction_id`, `token`, `amount` since each order can have multiple fees taken per trade.

I'm fine with that as well, though since I assume decentralizing the DB becomes easier the more straight forward our schemes are.

Reply:
> i just gave up

Neither `WHERE (order_uid, auction_id) IN ({})` nor `WHERE (order_uid, auction_id) = ANY($1)` worked out?

Reply:
This works: https://github.com/cowprotocol/services/commit/7d1b92c63f691217b745474827eccee6ea6a1884

Reply:
> AFAIU the alternative would be an extra table with ...

We could also have two columns with vectors of scalar types (varchar, bigint) to keep it in a single table